### PR TITLE
make apiservices.apiregistration.k8s.io discoverable in openapi/v3

### LINKED ISSUE #118878: apiregistration.k8s.io is not available in openapi/v3 endpoint and discovery
### What happened?

it is not possible to discover apiregistration.k8s.io in openapi/v3. 

```shell
$ kubectl explain apiservices
error: couldn't find resource for "apiregistration.k8s.io/v1, Resource=apiservices"
```

This was possible to do with v2.

```shell
$ kubectl explain apiservices --output=plaintext-openapiv2
KIND:     APIService
VERSION:  apiregistration.k8s.io/v1
...
```

### What did you expect to happen?

openapi/v3 discovery works for apiregistration.k8s.io 

### How can we reproduce it (as minimally and precisely as possible)?

```shell
$ curl -k -H "Authorization: Bearer $TOKEN" https://localhost:6443/openapi/v3 | grep apiregistration
```

```shell
$ curl -k -H "Authorization: Bearer $TOKEN" https://localhost:6443/openapi/v3 | grep apiservice
```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.2", GitCommit:"7f6f68fdabc4df88cfea2dcf9a19b2b830f1e647", GitTreeState:"archive", BuildDate:"2023-05-28T13:03:55Z", GoVersion:"go1.20.4", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"28+", GitVersion:"v1.28.0-alpha.3.69+ba6d2674ca6b7b-dirty", GitCommit:"ba6d2674ca6b7b4992189bba1516062004f05ed3", GitTreeState:"dirty", BuildDate:"2023-06-25T23:36:20Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
NONE
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by atiratree**
#### What type of PR is this?

/kind bug
/kind regression
/sig api-machinery

#### What this PR does / why we need it:
make apiservices.apiregistration.k8s.io discoverable in openapi/v3

#### Which issue(s) this PR fixes:

Fixes #118878

#### Does this PR introduce a user-facing change?

```release-note
Fixes a regression exposed in 1.27 by kubectl switching to openapi v3 by making apiregistration.k8s.io discoverable in openapi/v3 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by atiratree** [comment]
side topic: do we want `kubectl get apiservices.apiregistration.k8s.io` to be self referential?  That is if we should include `v1.apiregistration.k8s.io` as a local service? Not sure if it is a good idea, just asking out of curiosity.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go [hunk 1]
```diff
@@ -73,29 +77,60 @@ func (s *specProxier) GetAPIServiceNames() []string {
...
+
+		for gv, ws := range grouped {
+			spec, err := builder3.BuildOpenAPISpecFromRoutes(restfuladapter.AdaptWebServices(ws), config)
+			if err != nil {
+				klog.Errorf("Failed to build OpenAPI v3 for group %s, %q", gv, err)
+
+			}
+			aggregatorWebServicesHandler.UpdateGroupVersion(gv, spec)
+		}
+		apiServiceName := fmt.Sprintf(localDelegateChainNamePattern, i)
```
Comment (Jefftree) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go:108]: This isn't really a delegation handler in the traditional sense. The aggregator serves the path, can we use a constant name for this to not affect the local delegation numbers?

**Review by Jefftree — COMMENTED** [review]
Thanks for bringing this up and fixing it. A couple small comments otherwise LGTM

**Comment by Jefftree** [comment]
Actually this is a duplicate of https://github.com/kubernetes/kubernetes/pull/118884

**Comment by alexzielenski** [comment]
wow @atiratree  I was reviewing your other PR for other resources that fail with `kubectl explain` and ended up fixing this independently. Looks like our fixes are the same, but you got a bit farther with writing tests. Feel free to take any implementation from my PR #118884. Will close in favor of this

**Comment by atiratree** [comment]
@alexzielenski ow, what a coincidence. Sorry about that. Will take a look at your PR.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator_test.go [hunk 2]
```diff
@@ -239,3 +290,20 @@ func sendReq(t *testing.T, handler http.Handler, path string) []byte {
 	handler.ServeHTTP(writer, req)
 	return writer.data
 }
+
+func testGetOpenAPIDefinitions(_ openapicommon.ReferenceCallback) map[string]openapicommon.OpenAPIDefinition {
```
[the remark this answers is not in the document]
Comment (atiratree) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator_test.go]: using APIService prefix now

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go (hunk 1 above)
Comment (atiratree) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go]: renamed to `k8s_internal_local_kube_aggregator_types` to at least have the same prefix as the delegation handlers

**Comment by atiratree** [comment]
@Jefftree updated the PR, using @alexzielenski simpler implementation now

**Comment by atiratree** [comment]
I have opened a cherry-pick into 1.27 https://github.com/kubernetes/kubernetes/pull/119841. Not sure if we should backport this further back, but we should at least do 1.27, since the OpenAPIV3 feature has been graduated to stable.

**Pull request merged**